### PR TITLE
Add JSON reader options structs to pylibcudf

### PR DESCRIPTION
**Pull request by Matt711**
## Description

Apart of #17565 
## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: python/pylibcudf/pylibcudf/io/json.pxd [hunk 1]
```diff
@@ -6,23 +6,53 @@ from pylibcudf.io.types cimport (
...
-    bool mixed_types_as_string = *,
-    bool prune_columns = *,
-    json_recovery_mode_t recovery_mode = *,
-    dict extra_parameters = *,
-)
+cdef class JsonReaderOptions:
+    cdef json_reader_options c_obj
+    cdef SourceInfo source
+    # cpdef void set_dtypes_list(self, list types)
+    # cpdef void set_dtypes_map(self, dict types)
```
Comment (Matt711) [inline python/pylibcudf/pylibcudf/io/json.pxd:21]:
```suggestion
```

File: python/pylibcudf/pylibcudf/io/json.pyx [hunk 2]
```diff
@@ -73,88 +77,358 @@ cdef json_reader_options _setup_json_reader_options(
...
+
+        Returns
+        -------
+        None
+        """
+        cdef vector[data_type] types_vec
+        if isinstance(types[0], tuple):
+            self.c_obj.set_dtypes(_generate_schema_map(types))
+        else:
+            for dtype in types:
```
Comment (mroeschke) [inline python/pylibcudf/pylibcudf/io/json.pyx:167]: Should `types_vec.reserve(len(types))` beforehand?

Comment (Matt711) [reply python/pylibcudf/pylibcudf/io/json.pyx:167]: I'm not sure how much more performant it would be in this case (simple loop). But I'll add it

Comment (Matt711) [inline python/pylibcudf/pylibcudf/io/json.pyx:167]:
```suggestion
            types_vec.reserve(len(types))
            for dtype in types:
```

Comment (Matt711) [reply python/pylibcudf/pylibcudf/io/json.pyx:167]: It'll depend on the number of reallocations done inside the loop (which depends on len(types) and the capacity of the vector).

**Review by mroeschke — APPROVED** [review]

**Pull request merged**